### DISCUSSION TITLE
`sparkStore()` or `sparkStoreFromClass()` ?

### DISCUSSION THREAD

**Discussion by khalwat** [putyourlightson/craft-spark#8] [Ideas]
So I have a fairly nerdy idea or possible PR for you if you like it... `sparkStore()` or `sparkStoreFromClass()` if you want to be verbose... you pass in the class name, and it'll generate the JSON store that represents the class's properties.

So, for people like me who will be using `Attributes` to define the store data structure in order to get autocomplete of `store.` in my Twig templates, that class can be the source of truth. So in Twig it'd look like this: 
```twig
        <section data-store="{{ sparkStore('attributes\\stores\\AutocompleteSearch') }}">
```
and if the class looks like this:
```php
<?php

namespace attributes\stores;

use Attribute;
use putyourlightson\spark\models\StoreModel;

#[Attribute]
class AutocompleteSearch extends StoreModel
{
    /** @var string $search The string to search for */
    public string $search;
}
```
the output would be this:
```twig
        <section data-store="{ search: ''}">
```

This gets us nice autocomplete in our Twig templates if we do this:
```twig
{# @var store \attributes\stores\AutocompleteSearch #}
```

<img width="736" alt="Screenshot 2024-10-20 at 20 35 25" src="https://github.com/user-attachments/assets/ead787a0-c1ba-44a7-af47-c8eccac7e2ea">

And then for completeness, we should let Composer's autoload know about our class in our `composer.json`:

```json
  "autoload": {
    "psr-4": {
      "attributes\\": "src/attributes/"
    }
  },
```

...and it means the `AutocompleteSearch` Attributes model is the source of truth for the store data structure. We only change it there in one place, rather than needing to edit it both in the PHP class and then also in the Twig templates when we change it.

Thoughts? I realize it's a little esoteric, but people can ignore it if they don't want to use it.

**Comment by bencroker** [comment]
You know I’m a sucker for autocomplete, so I like it! 

I already was considering adding something like `sparkStore()` that accepts an array of key-value pairs and returns a JSON encoded string of JS-safe (escaped) values, so this could be a nice addition. I’d probably name it `sparkStoreFromClass()`, as you suggest, and use reflection to extract the class’s public properties. Is that along the lines of what you had in mind?

**Comment by khalwat** [comment]
Yessir, you got it! Can even pull out the default values (if any).

**Comment by khalwat** [comment]
Would `sparkStore()` be doing anything that the baked in `json_encode` filter doesn't already do?

**Reply by bencroker** [reply]
Just some validation. I thought it might make sense to escape values, as per @croxton ’s code in https://github.com/putyourlightson/craft-spark/discussions/10#discussion-7347679, but I think that was only required there since store values were being set from query string params (unescaped user input).
https://github.com/putyourlightson/craft-spark/commit/9988b2680667856e02c85876f07d75d4d678bf0a#diff-1f4c279925e699c9acb4d01b87125110ce0f40c403a2abbccdd80acc819ec262R89

**Reply by bencroker** [reply]
Added some docs the new methods under https://putyourlightson.com/plugins/spark#data-store

**Comment by mattbloomfield** [comment]
Ah yes, love this. Great idea!

**Comment by khalwat** [comment]
Can you make the `sparkStoreFromClass()` work recursively, so that nested arrays/objects work? The Datastar store supports nested JSON values, and I'm planning to use a nested object in my demo 🙏 

https://github.com/khalwat/spark-datastar-pokemon-demo/commit/443033bbc7e838028c86efd117279cf853638148

**Reply by bencroker** [reply]
Done and released in 0.0.9.
